_drafts/2020-12-14-data_science_interview_prep_data_manipulation.md

```diff
@@ -0,0 +1,378 @@
+FROM collisions
+WHERE motorcycle_collision = True
+  AND party_count = 1
+GROUP BY collision_year
+ORDER BY collision_year
+```
+
+This gives us:
+
+<div class="low-width-table" markdown="1" style="max-width: 20%">
+
+|   collision_year |   collision_count |
+|-----------------:|------------------:|
+|             2001 |              3258 |
+|             2002 |              3393 |
+|             2003 |              3822 |
+|             2004 |              3955 |
+|             2005 |              3755 |
+|             2006 |              3967 |
+|             2007 |              4513 |
+|             2008 |              4948 |
+|             2009 |              4266 |
+|             2010 |              3902 |
+|             2011 |              4054 |
+|             2012 |              4143 |
+|             2013 |              4209 |
+|             2014 |              4267 |
+|             2015 |              4415 |
+|             2016 |              4471 |
+|             2017 |              4373 |
+|             2018 |              4240 |
+|             2019 |              3772 |
+|             2020 |              2984 |
+
+</div>
+
+The count is low in 2020 primarily because the data doesn't cover the whole
+year. It is also low due to the COVID pandemic which kept people off the
+streets, at least initially. To differentiate these two causes we could
+compare month by month to last year.
+
+</div>
+
+### What make of vehicle has the largest fraction of accidents on the weekend? During the work week?
+
+Weekdays are generally commute and work-related traffic, while weekends
+involves recreational travel. Do we see different vehicles involved in
+collisions on these days?
+
+Only consider vehicle makes with at least 10,000 collisions.
+
+<button id="button" onclick="showhide(hidden4)">Show solution</button>
+<div class="hidden" id="hidden4" markdown="1" style="display: none;">
+
+This query is tricky. We need to aggregate collisions by vehicle make, which
+means we need the parties table. We also care about when the crash happened,
+which means we need the collisions table. So we need to join these tables
+together.
+
+I use a sub-query to do the aggregation. A `WTIH` clause keeps it tidy so we
+don't have to copy/paste the sub-query twice. I use `HAVING` to filter out
+makes with too few collisions; it has to be `HAVING` and not `WHERE` because
+it filters **after** the aggregation.
+
+I then construct two queries that read from the sub-query to select the
+highest row for the weekend and weekdays. I `UNION` to two queries together so
+we end up with a single table containing our results. The double select is to
+allow the `ORDER BY` before the `UNION`.
+
+In an interview setting, I would have just written two simpler queries: one
+that gets the highest weekend fraction and one that gets the highest weekday
+fraction with a lot of copy and pasted code. This would have been a lot faster
+to come up with and write.
```

Comment:
Would it make sense to show these two simpler queries? Maybe discuss how one can answer an interview using the two simple queries then discuss how they can be combined into a single query? (are there advantages or drawbacks to that approach?)

Reply:
Great feedback!